File: 1769.minimum-number-of-operations-to-move-all-balls-to-each-box.cpp
/*
 * @lc app=leetcode id=1769 lang=cpp
 *
 * [1769] Minimum Number of Operations to Move All Balls to Each Box
 */

// @lc code=start
#include<vector>
#include<queue>
#include<string>
using std::queue;
using std::string;
using std::vector;
class Solution {
public:
    vector<int> minOperations(string boxes) {
        queue<int> dq {};
        int leftS = 0, rightS = 0, n = boxes.size(), leftCount = 0;
        vector<int> result {};

        for (int i = 0; i < n; i++) {
            if (boxes[i] == '1') {
                dq.push(i);
                rightS += i;
            }
        }

        for (int j = 0; j < n; j++) {
            if (dq.size() > 0 && j == dq.front()) {
                leftS += dq.front();
                dq.pop();
                rightS -= j;
                leftCount++;
            }

            result.push_back(leftCount * j - leftS + rightS - dq.size() * j);
        }

        return result;        
    }
};
// @lc code=end

File: 1769.minimum-number-of-operations-to-move-all-balls-to-each-box.cs
/*
 * @lc app=leetcode id=1769 lang=csharp
 *
 * [1769] Minimum Number of Operations to Move All Balls to Each Box
 */

// @lc code=start
using System.Collections.Generic;

public class Solution {
    public int[] MinOperations(string boxes) {
        Queue<int> dq = new Queue<int>();
        int leftS = 0, rightS = 0, n = boxes.Length, leftCount = 0;
        List<int> result = new List<int>();

        for (int i = 0; i < n; i++) {
            if (boxes[i] == '1') {
                dq.Enqueue(i);
                rightS += i;
            }
        }

        for (int j = 0; j < n; j++) {
            if (dq.Count > 0 && j == dq.Peek()) {
                leftS += dq.Dequeue();
                rightS -= j;
                leftCount++;
            }

            result.Add(leftCount * j - leftS + rightS - dq.Count * j);
        }

        return result.ToArray();
    }
}

// @lc code=end

